### ISSUE TITLE
How to avoid dag_direct_scheduler warning "Detected a requirement ... that accesses uninitialized data"?

### ISSUE THREAD

**Issue by mabraham** [AdaptiveCpp/AdaptiveCpp#609]
GROMACS is using hipSYCL to call rocFFT to do 3D FFTs, first forward, then backward. The ability to enqueue a custom operation has been very useful for that, so thanks!

However, regardless of the build configuration of either GROMACS or hipSYCL, I get a run-time warning:
`[hipSYCL Warning] dag_direct_scheduler: Detected a requirement that is neither of discard access mode (SYCL 1.2.1) nor noinit property (SYCL 2020) that accesses uninitialized data. Consider changing to discard/noinit. Optimizing potential data transfers away.`

The code triggering that warning is https://gitlab.com/gromacs/gromacs/-/merge_requests/1833/diffs#c95806302062c303359425a19287ab63d40b9a16_0_342, where `impl_->realGrid_` is filled with data and `impl_->complexGrid_` is allocated but not filled. If instead I declare like `auto complexGridAccessor = cl::sycl::accessor<float, 1, mode::write>(*impl_->complexGrid_, cgh, cl::sycl::property::no_init{});` the warning remains. If I fill the complex grid before this kernel is called, then the warning goes away, but as this buffer is strictly write-only, I should not need to do that.

How should we write the code to avoid generating such a warning?

One can suppress it by setting`HIPSYCL_DEBUG_WARNING` appropriately, but I think it would be better to fix in code somewhere.

I can produce a minimal reproducer if that would be useful.

**Comment by fodinabor** [comment]
Hi,
I'd say the `no_init` property should help there. Any chance you could check with PR #606, it seems to address misinterpreting the `no_init` property?

**Comment by mabraham** [comment]
Yes, I checked the code of the branch of #606 (commit a44be5409052fba22443f32d45156085dc92120d). All three versions of the above GROMACS code segfault, but at least now the version using the `no_init` property did not give the warning.

The first call of `perform3dFFt` to do the real-to-complex transform works fine, but the segfault occurs when the test case calls it a second time to do the reverse transformation on the same buffers (after clearing the realGrid which is now to become the output buffer, see https://gitlab.com/gromacs/gromacs/-/merge_requests/1833/diffs#dcdf5328e81184fc1a49bab6c47371cc8c059104_435_439).

Backtrace of segfault:
```
(lldb) bt
error: need to add support for DW_TAG_base_type 'auto' encoded with DW_ATE = 0x0, bit_size = 0
* thread #1, name = 'fft-test', stop reason = signal SIGSEGV: invalid address (fault address: 0x5a)
  * frame #0: 0x00007ffff6dd580b libgromacs.so.7`void hipsycl::sycl::accessor<float, 1, (hipsycl::sycl::access_mode)1, (hipsycl::sycl::target)0, (hipsycl::sycl::accessor_variant)0>::bind_to_buffer<hipsycl::sycl::buffer<float, 1, std::allocator<float> > >(hipsycl::sycl::buffer<float, 1, std::allocator<float> >&, hipsycl::sycl::id<1>, hipsycl::sycl::range<1>) [inlined] std::__shared_ptr<hipsycl::rt::data_region<void*>, (__gnu_cxx::_Lock_policy)2>::__shared_ptr(this=<unavailable>, (null)=0x000000000000005a)2> const&) at shared_ptr_base.h:1167:7
    frame #1: 0x00007ffff6dd580b libgromacs.so.7`void hipsycl::sycl::accessor<float, 1, (hipsycl::sycl::access_mode)1, (hipsycl::sycl::target)0, (hipsycl::sycl::accessor_variant)0>::bind_to_buffer<hipsycl::sycl::buffer<float, 1, std::allocator<float> > >(hipsycl::sycl::buffer<float, 1, std::allocator<float> >&, hipsycl::sycl::id<1>, hipsycl::sycl::range<1>) [inlined] std::shared_ptr<hipsycl::rt::data_region<void*> >::shared_ptr(this=<unavailable>, (null)=nullptr) at shared_ptr.h:129
    frame #2: 0x00007ffff6dd580b libgromacs.so.7`void hipsycl::sycl::accessor<float, 1, (hipsycl::sycl::access_mode)1, (hipsycl::sycl::target)0, (hipsycl::sycl::accessor_variant)0>::bind_to_buffer<hipsycl::sycl::buffer<float, 1, std::allocator<float> > >(hipsycl::sycl::buffer<float, 1, std::allocator<float> >&, hipsycl::sycl::id<1>, hipsycl::sycl::range<1>) [inlined] std::shared_ptr<hipsycl::rt::data_region<void*> > hipsycl::sycl::detail::extract_buffer_data_region<hipsycl::sycl::buffer<float, 1, std::allocator<float> > >(buff=0x0000000000a156a0) at buffer.hpp:1154
    frame #3: 0x00007ffff6dd5807 libgromacs.so.7`void hipsycl::sycl::accessor<float, 1, (hipsycl::sycl::access_mode)1, (hipsycl::sycl::target)0, (hipsycl::sycl::accessor_variant)0>::bind_to_buffer<hipsycl::sycl::buffer<float, 1, std::allocator<float> > >(this=0x00007fffffffbf80, buff=0x0000000000a156a0, accessOffset=id<1> @ r13, accessRange=range<1> @ r12) at accessor.hpp:1082
    frame #4: 0x00007ffff6dd56a6 libgromacs.so.7`void hipsycl::sycl::accessor<float, 1, (hipsycl::sycl::access_mode)1, (hipsycl::sycl::target)0, (hipsycl::sycl::accessor_variant)0>::init<hipsycl::sycl::buffer<float, 1, std::allocator<float> > >(this=0x00007fffffffbf80, buff=0x0000000000a156a0, cgh=0x00007fffffffc010, offset=id<1> @ r12, access_range=range<1> @ r14, prop_list=<unavailable>) at accessor.hpp:1031:5
    frame #5: 0x00007ffff6dcfbe4 libgromacs.so.7`gmx::Gpu3dFft::perform3dFft(gmx_fft_direction, void**) [inlined] void hipsycl::sycl::accessor<float, 1, (hipsycl::sycl::access_mode)1, (hipsycl::sycl::target)0, (hipsycl::sycl::accessor_variant)0>::init<hipsycl::sycl::buffer<float, 1, std::allocator<float> > >(this=0x00007fffffffbf80, buff=0x0000000000a156a0, cgh=0x00007fffffffc010, prop_list=0x00007fffffffbd10) at accessor.hpp:1039:5
    frame #6: 0x00007ffff6dcfbbe libgromacs.so.7`gmx::Gpu3dFft::perform3dFft(gmx_fft_direction, void**) [inlined] hipsycl::sycl::accessor<float, 1, (hipsycl::sycl::access_mode)1, (hipsycl::sycl::target)0, (hipsycl::sycl::accessor_variant)0>::accessor<std::allocator<float>, 1, (this=<unavailable>, bufferRef=<unavailable>, commandGroupHandlerRef=<unavailable>, prop_list=<unavailable>)0>(hipsycl::sycl::buffer<float, 1, std::allocator<float> >&, hipsycl::sycl::handler&, hipsycl::sycl::property_list const&) at accessor.hpp:617
    frame #7: 0x00007ffff6dcfba3 libgromacs.so.7`gmx::Gpu3dFft::perform3dFft(gmx_fft_direction, void**) [inlined] gmx::Gpu3dFft::perform3dFft(this=<unavailable>, cgh=0x00007fffffffc010)::'lambda0'(hipsycl::sycl::handler&)::operator()(hipsycl::sycl::handler&) const at gpu_3dfft_sycl_hipsycl.cpp:363
    frame #8: 0x00007ffff6dcfb8a libgromacs.so.7`gmx::Gpu3dFft::perform3dFft(gmx_fft_direction, void**) [inlined] std::shared_ptr<hipsycl::rt::dag_node> hipsycl::sycl::queue::execute_submission<gmx::Gpu3dFft::perform3dFft(this=0x00000000009fe4f8, cgf=(anonymous class) @ r14, cgh=0x00007fffffffc010)::'lambda0'(hipsycl::sycl::handler&)>(gmx::Gpu3dFft::perform3dFft(gmx_fft_direction, void**)::'lambda0'(hipsycl::sycl::handler&), hipsycl::sycl::handler&) at queue.hpp:872
    frame #9: 0x00007ffff6dcfaf4 libgromacs.so.7`gmx::Gpu3dFft::perform3dFft(gmx_fft_direction, void**) [inlined] hipsycl::sycl::event hipsycl::sycl::queue::submit<gmx::Gpu3dFft::perform3dFft(gmx_fft_direction, void**)::'lambda0'(hipsycl::sycl::handler&)>(this=0x00000000009fe4f8, prop_list=0x00007fffffffbcf0, cgf=(anonymous class) @ r14)::'lambda0'(hipsycl::sycl::handler&)) at queue.hpp:351
    frame #10: 0x00007ffff6dcfaf4 libgromacs.so.7`gmx::Gpu3dFft::perform3dFft(gmx_fft_direction, void**) [inlined] hipsycl::sycl::event hipsycl::sycl::queue::submit<gmx::Gpu3dFft::perform3dFft(gmx_fft_direction, void**)::'lambda0'(hipsycl::sycl::handler&)>(this=0x00000000009fe4f8, cgf=(anonymous class) @ r14)::'lambda0'(hipsycl::sycl::handler&)) at queue.hpp:359
    frame #11: 0x00007ffff6dcfaf4 libgromacs.so.7`gmx::Gpu3dFft::perform3dFft(this=0x00007fffffffc418, dir=<unavailable>, (null)=<unavailable>) at gpu_3dfft_sycl_hipsycl.cpp:360
    frame #12: 0x00000000004137b1 fft-test`gmx::test::FFTTest3D_GpuReal5_6_9_Test::TestBody(this=0x00000000009ff2a0) at fft.cpp:452:18
    frame #13: 0x00007ffff485ae84 libgtest.so.1.11.0`void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) [inlined] void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(object=<unavailable>, method=<unavailable>, location="the test body")(), char const*) at gtest.cc:2631:10
    frame #14: 0x00007ffff485ae72 libgtest.so.1.11.0`void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(object=<unavailable>, method=00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00, location="the test body")(), char const*) at gtest.cc:2667
    frame #15: 0x00007ffff4833b61 libgtest.so.1.11.0`testing::Test::Run(this=0x00000000009ff2a0) at gtest.cc:2706:5
    frame #16: 0x00007ffff4834e30 libgtest.so.1.11.0`testing::TestInfo::Run(this=0x0000000000845b20) at gtest.cc:2885:11
    frame #17: 0x00007ffff48356f7 libgtest.so.1.11.0`testing::TestSuite::Run(this=0x00000000008459d0) at gtest.cc:3039:28
    frame #18: 0x00007ffff4845677 libgtest.so.1.11.0`testing::internal::UnitTestImpl::RunAllTests(this=<unavailable>) at gtest.cc:5897:44
    frame #19: 0x00007ffff485bad4 libgtest.so.1.11.0`bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) [inlined] bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(object=<unavailable>, method=<unavailable>, location="auxiliary test code (environments or event listeners)")(), char const*) at gtest.cc:2631:10
    frame #20: 0x00007ffff485bac2 libgtest.so.1.11.0`bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(object=<unavailable>, method=00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00, location="auxiliary test code (environments or event listeners)")(), char const*) at gtest.cc:2667
    frame #21: 0x00007ffff48450a9 libgtest.so.1.11.0`testing::UnitTest::Run(this=0x00007ffff487ade0) at gtest.cc:5464:10
    frame #22: 0x000000000042b4db fft-test`main [inlined] RUN_ALL_TESTS() at gtest.h:2492:46
    frame #23: 0x000000000042b4ce fft-test`main(argc=1, argv=0x00007fffffffc838) at unittest_main.cpp:77
    frame #24: 0x00007ffff36a30b3 libc.so.6`__libc_start_main + 243
    frame #25: 0x0000000000411c3e fft-test`_start + 46
```
GROMACS code at the segfault
```
frame #7: 0x00007ffff6dcfba3 libgromacs.so.7`gmx::Gpu3dFft::perform3dFft(gmx_fft_direction, void**) [inlined] gmx::Gpu3dFft::perform3dFft(this=<unavailable>, cgh=0x00007fffffffc010)::'lambda0'(hipsycl::sycl::handler&)::operator()(hipsycl::sycl::handler&) const at gpu_3dfft_sycl_hipsycl.cpp:363
   360 	        impl_->queue_.submit([&](cl::sycl::handler& cgh) {
   361 	            using cl::sycl::access::mode;
   362 	            //auto realGridAccessor    = impl_->realGrid_->get_access<mode::write>(cgh);
-> 363 	            auto realGridAccessor = cl::sycl::accessor<float, 1, mode::write>(*impl_->realGrid_, cgh, cl::sycl::property::no_init{});
   364 	            auto complexGridAccessor = impl_->complexGrid_->get_access<mode::read>(cgh);
   365 	            cgh.hipSYCL_enqueue_custom_operation([=](cl::sycl::interop_handle& h) {
   366 	                void* realGrid    = h.get_native_mem<cl::sycl::backend::hip>(realGridAccessor);
```

The `error: need to add support for DW_TAG_base_type 'auto' encoded with DW_ATE = 0x0, bit_size = 0` suggests to me that the new feature might be incomplete?

**Comment by fodinabor** [comment]
The `error: need to add support for DW_TAG_base_type 'auto' encoded with DW_ATE = 0x0, bit_size = 0` seems to be a LLDB thing and should only impact the quality of the backtrace, I guess: https://bugs.llvm.org/show_bug.cgi?id=46775

So before merging the PR #606, it was successful, just emitted the warning? do you know the commit you worked with before?

Still not quite sure the segfault is related to the #606 change, though.. Could you test whether the access mode `discard_write` results in the same issue?

It looks like the data `shared_ptr` inside the `buffer_impl` has been destructed already, note the `this=<unavailable>` for the copy ctor in the top two frames:
```
  * frame #0: 0x00007ffff6dd580b libgromacs.so.7`void hipsycl::sycl::accessor<float, 1, (hipsycl::sycl::access_mode)1, (hipsycl::sycl::target)0, (hipsycl::sycl::accessor_variant)0>::bind_to_buffer<hipsycl::sycl::buffer<float, 1, std::allocator<float> > >(hipsycl::sycl::buffer<float, 1, std::allocator<float> >&, hipsycl::sycl::id<1>, hipsycl::sycl::range<1>) [inlined] std::__shared_ptr<hipsycl::rt::data_region<void*>, (__gnu_cxx::_Lock_policy)2>::__shared_ptr(this=<unavailable>, (null)=0x000000000000005a)2> const&) at shared_ptr_base.h:1167:7
    frame #1: 0x00007ffff6dd580b libgromacs.so.7`void hipsycl::sycl::accessor<float, 1, (hipsycl::sycl::access_mode)1, (hipsycl::sycl::target)0, (hipsycl::sycl::accessor_variant)0>::bind_to_buffer<hipsycl::sycl::buffer<float, 1, std::allocator<float> > >(hipsycl::sycl::buffer<float, 1, std::allocator<float> >&, hipsycl::sycl::id<1>, hipsycl::sycl::range<1>) [inlined] std::shared_ptr<hipsycl::rt::data_region<void*> >::shared_ptr(this=<unavailable>, (null)=nullptr) at shared_ptr.h:129
```

My best guess how that happens is that there's a `sycl::buffer` instance that is destructed twice, thus freeing the `shared_ptr<buffer_impl>`, which implicitly deletes the impl's `data` member.. so the `realGrid_` still holds a valid `buffer` copy, but the wrapped `buffer_impl` is already destructed.
How that happens? no clue...

**Comment by illuhad** [comment]
Hi Mark,

the warning you mentioned is a notification for performance portability; hipSYCL informs you that it has detected an access to a buffer that does not contain initialized data. SYCL has an optimization hint for such cases: The `no_init` accessor property (or `discard_write/discard_read_write` access mode in SYCL 1.2.1) informs SYCL that any buffer data content can be discarded. In this case a SYCL implementation can pretty much guarantee that no data transfers will take place. Using `no_init` to access uninitialized data is therefore good practice.
As the warning indicates, hipSYCL has still optimized data transfers away, but you might still want to use `no_init` for best practice, and because the behavior of other implementations might be different.

So, in summary, this warning is a recommendation regarding style and performance portability to other implementations. It does not have an effect on correctness or performance as long as you stay with hipSYCL.

Regarding the segfaults I would recommend to try if it reproduces when running on CPU with address sanitizers enabled.

**Comment by illuhad** [comment]
>  If instead I declare like auto complexGridAccessor = cl::sycl::accessor<float, 1, mode::write>(*impl_->complexGrid_, cgh, cl::sycl::property::no_init{}); the warning remains

~I don't understand how this behavior could have been changed by PR #606~. If there is a case of the warning remaining even though `no_init` was passed I would be interested in a reproducer.

**Comment by mabraham** [comment]
While attempting a reproducer, I made some new hipSYCL test cases based on #606, and observed with

```
BOOST_AUTO_TEST_CASE(get_access_simple) {
  namespace s = cl::sycl;

  std::vector<float> v(32, 1.0);
  s::buffer<float, 1> buf(32);

  s::queue queue;
  queue.submit([&](s::handler& cgh) {
                   auto acc_from_get_access = buf.get_access<s::access::mode::write>(cgh);
                   cgh.copy(v.data(), acc_from_get_access);
               });
}

BOOST_AUTO_TEST_CASE(constructor_simple) {
  namespace s = cl::sycl;

  std::vector<float> v(32, 1.0);
  s::buffer<float, 1> buf(32);

  s::queue queue;
  queue.submit([&](s::handler& cgh) {
                   auto acc_constructed = s::accessor { buf, cgh, s::write_only};
                   cgh.copy(v.data(), acc_constructed);
               });
}

BOOST_AUTO_TEST_CASE(constructor_no_init) {
  namespace s = cl::sycl;

  std::vector<float> v(32, 1.0);
  s::buffer<float, 1> buf(32);

  s::queue queue;
  queue.submit([&](s::handler& cgh) {
                   auto acc_constructed_no_init = s::accessor { buf, cgh, s::write_only, s::property::no_init{}};
                   cgh.copy(v.data(), acc_constructed_no_init);
               });
}

BOOST_AUTO_TEST_CASE(get_access_no_init) {
  namespace s = cl::sycl;

  std::vector<float> v(32, 1.0);
  s::buffer<float, 1> buf(32);

  s::queue queue;
  queue.submit([&](s::handler& cgh) {
                   auto acc_from_get_access_no_init = buf.get_access(cgh, s::write_only, s::property::no_init{});
                   cgh.copy(v.data(), acc_from_get_access_no_init);
               });
  queue.wait();
}
```
the output
```
$ ./sycl_tests --run_test=accessor_tests/get_access_no_init
Running 1 test case...
[hipSYCL Warning] dag_node: Destructor invoked before operation has completed, this should never happen.
[hipSYCL Error] Attempted to access accessor that was not registered with the handler
unknown location(0): fatal error: in "accessor_tests/get_access_no_init": unknown type
/nethome/mabraham/git/hipSYCL/tests/sycl/accessor.cpp(159): last checkpoint: "get_access_no_init" test entry

*** 1 failure is detected in the test module "hipsycl unit tests"
```
The warning is gone, but I expected that test to work.

I built the tests like
```
cmake ../tests -DCMAKE_BUILD_TYPE=Debug -DCMAKE_PREFIX_PATH=../install-feature-support-non-tagged-accessor-deduction-guides-relwithdebinfo\;${BOOST_HOME} -G Ninja -DCMAKE_CXX_COMPILER=/opt/tcbsys/rocm/4.2.0/llvm/bin/clang++ -DCMAKE_C_COMPILER=/opt/tcbsys/rocm/4.2.0/llvm/bin/clang -DHIPSYCL_TARGETS=hip:gfx906
```

**Comment by illuhad** [comment]
Thanks, I could reproduce. The issue is that the variadic `buffer::get_access()` takes its arguments by value. This causes the handler to be copied, and hence the accessor registers with a wrong handler (thank god current hipSYCL can diagnose that, otherwise the results would have been a lot of invalid memory accesses due to invalid accessors..).

The `get_access()` signature is exactly as defined in the spec, but IMHO it should take something like R-value references to avoid this issue: `template<typename... Args> auto get_access(Args&&... args)`.

With this change it runs as expected.

**Comment by mabraham** [comment]
Thanks, that fix looks good. I'll keep digging for further problems, as my original use case is not yet working well
